File: C#/src/Intermediate/day48/Exercise.cs
using System;
using System.Collections.Generic; 
using System.Text;

namespace Intermediate.day48
{
    public class Exercise
    {
        public static int[] ex1FrequencyOfElementQuery(int[] A, int[] B)
        {
            Dictionary<int, int> frequencyMap = new Dictionary<int, int>();
            foreach (int num in A)
            {
                if (frequencyMap.ContainsKey(num))
                {
                    frequencyMap[num]++;
                }
                else
                {
                    frequencyMap[num] = 1;
                }
            }

            int[] result = new int[B.Length];
            for (int i = 0; i < B.Length; i++)
            {
                result[i] = frequencyMap.ContainsKey(B[i]) ? frequencyMap[B[i]] : 0;
            }

            return result;
        }

        public static int ex2FirstRepeatingElement(int[] A)
        {
            HashSet<int> set = new HashSet<int>();
            foreach (int num in A)
            {
                if (set.Contains(num))
                {
                    return num;
                }
                else
                {
                    set.Add(num);
                }
            }

            return -1; // No repeating element found
        }

        public static int ex3SubArrayWith0Sum(int[] A)
        {
            HashSet<int> prefixSums = new HashSet<int>();
            int currentSum = 0;
            foreach (int num in A)
            {
                currentSum += num;
                if (currentSum == 0 || prefixSums.Contains(currentSum))
                {
                    return 1; // Subarray with sum 0 found
                }
                prefixSums.Add(currentSum);
            }
            return 0; // No subarray with sum 0 found
        }

        public static int ex5CountUniqueElements(int[] A)
        {
            HashSet<int> uniqueElements = new HashSet<int>();
            foreach (int num in A)
            {
                uniqueElements.Add(num);
            }
            return uniqueElements.Count;

        }
    }
}

File: Java/Intermediate/src/main/java/day48/Exercise.java
package day48;

import java.util.HashMap;
import java.util.HashSet;

public class Exercise {
    public static int[] ex1FrequencyOfElementQuery(int[] A, int[] B){
        HashMap<Integer, Integer> map = new HashMap<>();
        for(int i = 0; i < A.length; i++){
            if(map.containsKey(A[i])){
                map.put(A[i], map.get(A[i]) + 1);
            }
            else{
                map.put(A[i], 1);
            }
        }
        int[] result = new int[B.length];
        for(int i = 0; i < B.length; i++){
            if(map.containsKey(B[i])){
                result[i] = map.get(B[i]);
            }
            else{
                result[i] = 0;
            }
        }
        return result;
    }
    public static int ex2FirstRepeatingElement(int[] A){
        HashMap<Integer, Integer> map = new HashMap<>();
        for(int i = A.length-1; i >= 0; i--){
            if(map.containsKey(A[i])){
                return A[i];
            }
            else{
                map.put(A[i], 1);
            }
        }
        return -1;
    }
    public static int ex3SubArrayWith0Sum(int[] A){
        HashSet<Long> set = new HashSet<>();
        Long fs = new Long(0);
        for(int i = 0; i < A.length; i++){
            fs += A[i];
            if(set.contains(fs) || fs == 0){
                return i;
            }
            else{
                set.add(fs);
            }
        }
        return -1;
    }
    public static int ex5CountUniqueElements(int[] A){
        HashSet<Integer> set = new HashSet<>();
        for(int i = 0; i < A.length; i++){
            set.add(A[i]);
        }
        return set.size();
    }
}
